### PR TITLE
Add L2, H(curl) and H(div) errors for vector variables

### PR DESCRIPTION
**Pull request by nmnobre**
## Motivation
MOOSE supports vector variables for vector equations, but the postprocessing of the errors (to another solution, for example analytic in MMS) for these vector variables has not been created yet.
The equivalent PP for regular variable do exist

## Design 
Create postprocessors that use the variable norms for L2, Hdiv Hcurl  spaces for vector variables

## Impact
Facilitate analysis of solutions using vector variables

Hi,

I've double checked the errors for the two tests match those for the corresponding libMesh examples (vector_fe_ex3 and vector_fe_ex6).

Cheers,
-Nuno

### CODE REVIEW

File: test/tests/kernels/vector_fe/gold/coupled_electrostatics_hex27.e
Comment (GiudGiud) [inline test/tests/kernels/vector_fe/gold/coupled_electrostatics_hex27.e:1]: if you are testing a postprocessor, better to use a CSVDiff than an exodiff

Comment (GiudGiud) [reply test/tests/kernels/vector_fe/gold/coupled_electrostatics_hex27.e:1]: also better to add a new test to just test the new postprocessor than to regold 7 tests

**Comment by nmnobre** [comment]
Hey @GiudGiud,

Can you help me with the seg fault first?

The problem is the test only defines `vec_x` and `vec_y`:

https://github.com/idaholab/moose/blob/8379791d1be276b1663b45331a19612f79b7efb2/test/tests/postprocessors/element_vec_l2_error_pps/element_vec_l2_error.i#L126-L133

And, then, `_uz` is set to `nullptr` here:

https://github.com/farscape-project/moose/blob/3acf7cb294f4467d13b9598f61531fc833a367de/framework/src/postprocessors/ElementVectorL2Error.C#L44

But why? Why is `isParamValid("var_z")` returning `false`?
For instance, `isParamValid("function_z")` returns `true`, even though the test doesn't set `function_z`...

I can workaround this by checking if the pointer is null just as I'm about to dereference it, but I was hoping by setting default values I could allow the user to only set some of the variables.

**Comment by GiudGiud** [comment]
I m not sure.
But checking for isParamValid when you provide a default (so it should always be valid) is not useful either so you might as well just remove the problem by removing that call

**Comment by lindsayad** [comment]
Behind the scenes `params.addCoupledVar` creates a parameter of type `std::vector<VariableName>`. So from the point of view that the parameter is a collection of strings, supplying real numbers does not make the param valid. Somebody might reasonably assume that if they check `isParamValid("var_z")` and it's `true` that this means retrieving the parameter value will give them string(s).  And in fact `FEProblemBase::addObjectParamsHelper` assumes just that. I tried setting `_valid = true` in the parameters when a default value is supplied for a coupled variable and then immediately encountered an error when running your input because `FEProblemBase::addObjectParamsHelper` choked when trying to read what it thought would be a string from the "valid" `variable` parameter.

Now is there a discrepancy in the behavior between `isParamValid("coupled_var_name_with_default")` and `isParamValid("function_name_with_default")`? Yes there is ... but no one has complained about the latter yet.

**Comment by nmnobre** [comment]
> Behind the scenes `params.addCoupledVar` creates a parameter of type `std::vector<VariableName>`. So from the point of view that the parameter is a collection of strings, supplying real numbers does not make the param valid.

But I don't have any other choice, all `addCoupledVar` methods require reals...

> then immediately encountered an error when running your input because `FEProblemBase::addObjectParamsHelper` choked when trying to read what it thought would be a string from the "valid" `variable` parameter.

What's this method used for? The code I'm expanding here already provided reals as default values, and it worked!

File: framework/src/postprocessors/ElementVectorL2Error.C [hunk 1]
```diff
@@ -48,17 +71,15 @@ ElementVectorL2Error::getValue() const
...
-  sol_val(0) = _u[_qp];                          // required variable
-  func_val(0) = _funcx.value(_t, _q_point[_qp]); // required function
-
-  sol_val(1) = _v[_qp];
-  sol_val(2) = _w[_qp];
+  RealVectorValue func_val = isParamSetByUser("function")
+                                 ? _func.vectorValue(_t, _q_point[_qp])
+                                 : RealVectorValue(_funcx->value(_t, _q_point[_qp]),
+                                                   _funcy->value(_t, _q_point[_qp]),
+                                                   _funcz->value(_t, _q_point[_qp]));
```
Comment (lindsayad) [inline framework/src/postprocessors/ElementVectorL2Error.C:78]:
```suggestion
                                 : RealVectorValue(_funcx.value(_t, _q_point[_qp]),
                                                   _funcy.value(_t, _q_point[_qp]),
                                                   _funcz.value(_t, _q_point[_qp]));
```

File: framework/include/postprocessors/ElementVectorL2Error.h [hunk 2]
```diff
@@ -25,11 +30,15 @@ class ElementVectorL2Error : public ElementIntegralPostprocessor
...
-  const Function & _funcz;
-
-  const VariableValue & _u; // FE solution in x
-  const VariableValue & _v; // FE solution in y
-  const VariableValue & _w; // FE solution in z
+  /// vector or component-wise analytical solution to compare against
+  const Function & _func;
+  const Function * const _funcx;
+  const Function * const _funcy;
+  const Function * const _funcz;
```
Comment (lindsayad) [inline framework/include/postprocessors/ElementVectorL2Error.h:37]:
```suggestion
  const Function & _funcx;
  const Function & _funcy;
  const Function & _funcz;
```

File: framework/src/utils/InputParameters.C [hunk 3]
```diff
@@ -1090,6 +1090,13 @@ InputParameters::addDeprecatedParam<std::vector<MooseEnum>>(
              "using addDeprecatedParam, even if the parameter is not required!");
 }
 
+template <>
+void
+InputParameters::setParamHelper<std::vector<VariableName>, Real>(
+    const std::string & /*name*/, std::vector<VariableName> & /*l_value*/, const Real & /*r_value*/)
+{
+}
+
```
Comment (lindsayad) [inline framework/src/utils/InputParameters.C:1099]: Don't think we need this

File: framework/src/utils/InputParameters.C [hunk 4]
```diff
@@ -218,7 +218,7 @@ InputParameters::setDeprecatedVarDocString(const std::string & new_name,
 void
 InputParameters::addCoupledVar(const std::string & name, Real value, const std::string & doc_string)
 {
-  addParam<std::vector<VariableName>>(name, doc_string);
+  addParam<std::vector<VariableName>, Real>(name, value, doc_string);
```
Comment (lindsayad) [inline framework/src/utils/InputParameters.C:221]: Don't think we need this change

File: modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C [hunk 5]
```diff
@@ -53,7 +53,6 @@ PorousFlowSinglePhaseBase::validParams()
       "The name of the user object for fluid "
       "properties. Only needed if fluid_properties_type = PorousFlowSingleComponentFluid");
   params.addCoupledVar("mass_fraction_vars",
-                       {},
```
Comment (lindsayad) [inline modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C:56]: Don't think we need this change

**Review by lindsayad — CHANGES_REQUESTED** [review]
As implied by an earlier comment from @GiudGiud I think we are trying to solve a problem that we unnecessarily manufactured. Since you supply a default value, you can just use the `getFunction` and `coupledValue` APIs. I think we can drop the commit with the `InputParameters` changes

**Comment by nmnobre** [comment]
> As implied by an earlier comment from @GiudGiud I think we are trying to solve a problem that we unnecessarily manufactured. Since you supply a default value, you can just use the `getFunction` and `coupledValue` APIs. I think we can drop the commit with the `InputParameters` changes

Can't do that because I inherit from this class and those derived classes don't have the same parameters.

The last commit has solved the problem anyway, the only thing left is the EXODIFF checks in parallel that fail by a minute fraction, I'm just trying to tweak the solver until they converge to the same thing.

**Comment by lindsayad** [comment]
It's definitely non-standard practice to inherit from a class and not inherit its parameters. I would discourage that

**Comment by lindsayad** [comment]
If you inherit the parameters, it's perfectly valid to mark a formerly non-required parameter as required in the derived class. The result will be what you expect: the parameter will be required

**Comment by nmnobre** [comment]
> It's definitely non-standard practice to inherit from a class and not inherit its parameters. I would discourage that

It was much harder to do it the way I've done it, I'll give you that.
But I thought it was way more elegant to have the error classes inherit from each other, reflecting their mathematical nature.
My problem was that `ElementVectorL2Error` existed but didn't do what I wanted it to do (no vector variables), so I tried to keep its existing functionality for backwards compatibility. Of course I could have created a `ElementProperVectorL2Error`, but then whichever name I choose, it will always be a problem to explain the user that one uses a component-wise vector and the other a true vector variable.

**Comment by lindsayad** [comment]
My suggestion would be to leave the current class alone, and then create a new class that takes a vector functor. If you haven't worked with functors before, the documentation page is [here](https://mooseframework.inl.gov/syntax/Functors/index.html). Then all in one your approximate solution parameter would take vector variables, vector functions, and vector material properties. Similarly for your exact solution parameter (although you may only ever pass a vector function)

**Comment by lindsayad** [comment]
Although that would mean needing to add `curl` and `div` interfaces to the `MooseFunctor` hierarchy ... that is quite a bit of work

**Comment by lindsayad** [comment]
> If you inherit the parameters, it's perfectly valid to mark a formerly non-required parameter as required in the derived class. The result will be what you expect: the parameter will be required

Is there any issue with you doing this?

**Comment by nmnobre** [comment]
> > If you inherit the parameters, it's perfectly valid to mark a formerly non-required parameter as required in the derived class. The result will be what you expect: the parameter will be required
> 
> Is there any issue with you doing this?

What I'd need to do is extend the implementation of `HDivSemiError` and `HCurlSemiError` for component-wise specifications, at least for the variables. I guess I could leave the requirement for a vector-valued `Function` implementing `div` and `curl`, but lift said requirement on the variable... I can build the divergence and the curl from the gradient of the scalar variables...

But, hear me out, I thought we agree that the current implementation for `addCoupledVar` is deficient, it doesn't set the parameter as valid as one would expect when specifying the default, how's the proposed fix a problem?

**Comment by lindsayad** [comment]
> But, hear me out, I thought we agree that the current implementation for `addCoupledVar` is deficient, it doesn't set the parameter as valid as one would expect when specifying the default, how's the proposed fix a problem?

No, I don't agree. I don't think a numeric string is a valid value for a variable name. Someone could query a system for that variable name which certainly doesn't exist

**Comment by lindsayad** [comment]
> What I'd need to do is extend the implementation of `HDivSemiError` and `HCurlSemiError` for component-wise specifications, at least for the variables.

I don't think so. I would just do this
```
diff --git a/framework/src/postprocessors/ElementHCurlSemiError.C b/framework/src/postprocessors/ElementHCurlSemiError.C
index 8500983c25..5f74c59c09 100644
--- a/framework/src/postprocessors/ElementHCurlSemiError.C
+++ b/framework/src/postprocessors/ElementHCurlSemiError.C
@@ -15,12 +15,18 @@ registerMooseObject("MooseApp", ElementHCurlSemiError);
 InputParameters
 ElementHCurlSemiError::validParams()
 {
-  InputParameters params = ElementIntegralPostprocessor::validParams();
+  InputParameters params = ElementVectorL2Error::validParams();
   params.addClassDescription("Returns the H(curl)-seminorm of the difference between a pair of "
                              "computed and analytical vector-valued solutions.");
   params.addRequiredParam<FunctionName>("function",
                                         "The vector analytical solution to compare against");
   params.addRequiredCoupledVar("variable", "The vector FE solution");
+  params.suppressParameter<FunctionName>("function_x");
+  params.suppressParameter<FunctionName>("function_y");
+  params.suppressParameter<FunctionName>("function_z");
+  params.suppressParameter<std::vector<VariableName>>("var_x");
+  params.suppressParameter<std::vector<VariableName>>("var_y");
+  params.suppressParameter<std::vector<VariableName>>("var_z");
   return params;
 }
 
diff --git a/framework/src/postprocessors/ElementHDivSemiError.C b/framework/src/postprocessors/ElementHDivSemiError.C
index 037adbf5a4..7f5da8026e 100644
--- a/framework/src/postprocessors/ElementHDivSemiError.C
+++ b/framework/src/postprocessors/ElementHDivSemiError.C
@@ -15,12 +15,18 @@ registerMooseObject("MooseApp", ElementHDivSemiError);
 InputParameters
 ElementHDivSemiError::validParams()
 {
-  InputParameters params = ElementIntegralPostprocessor::validParams();
+  InputParameters params = ElementVectorL2Error::validParams();
   params.addClassDescription("Returns the H(div)-seminorm of the difference between a pair of "
                              "computed and analytical vector-valued solutions.");
   params.addRequiredParam<FunctionName>("function",
                                         "The vector analytical solution to compare against");
   params.addRequiredCoupledVar("variable", "The vector FE solution");
+  params.suppressParameter<FunctionName>("function_x");
+  params.suppressParameter<FunctionName>("function_y");
+  params.suppressParameter<FunctionName>("function_z");
+  params.suppressParameter<std::vector<VariableName>>("var_x");
+  params.suppressParameter<std::vector<VariableName>>("var_y");
+  params.suppressParameter<std::vector<VariableName>>("var_z");
   return params;
 }
 
```
If we are to follow the logic that

> it was way more elegant to have the error classes inherit from each other, reflecting their mathematical nature.

then I think this is the right way to go. I think we should either have no inheritance at all or follow standard MOOSE practice which is to inherit the parameters of the base class and then modify them as appropriate for the derived class. Then you are much more likely to be able to construct your base class in a standard way

**Comment by nmnobre** [comment]
Oh.... I didn't know you could suppress parameters... that might just be what we need then. :)

**Comment by lindsayad** [comment]
> Oh.... I didn't know you could suppress parameters... that might just be what we need then. :) That was indeed the purpose of all this trickery... suppressing the unused parameters. Thanks, I'll do that later. 🎉

🎉 

> PS: Next time I might call you on Slack so we can get to the answer a bit more quickly. 😝

Haha, sounds good. You can call any time!

**Comment by nmnobre** [comment]
All done.

@GiudGiud I've still regolded the RAVIART_THOMAS Exodiff tests as they weren't really all that well converged. I've separated the preprocessor tests into seven new CSVDiff tests as suggested though. 🙂

Thanks for you time and suggestions gents,
Have a very Merry Christmas 🎄

**Comment by lindsayad** [comment]
Hey Nuno, if I don't get to it during the INL holiday, I will be back to regular hours next Tuesday to review this

**Comment by nmnobre** [comment]
No worries, enjoy your holidays Alex!

File: framework/doc/content/source/postprocessors/ElementHCurlSemiError.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+\end{equation}
+
+where $||\cdot||$ denotes the l2-norm in $\mathbb{R}^d$.
+
+The nonlinear vector variable specified for $\vec{u}$ must be of the
+`NEDELEC_ONE` family, as MOOSE currently does not compute the curl for
+any other finite element family.
+
+$\vec{f}$ must be a [Function](syntax/Functions/index.md) object defining both
+the `vectorValue` and `curl` methods, e.g. `MooseParsedVectorFunction.md`.
```
Comment (lindsayad) [inline framework/doc/content/source/postprocessors/ElementHCurlSemiError.md:21]:
```suggestion
the `vectorValue` and `curl` methods, e.g. [MooseParsedVectorFunction.md].
```

File: framework/doc/content/source/postprocessors/ElementHDivSemiError.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+\begin{equation}
+    \left( \int_{\Omega} | \nabla \cdot \vec{u}  - \nabla \cdot \vec{f} |^2 \,\text{d}x \right) ^{1/2}
+\end{equation}
+
+where $|\cdot|$ denotes the absolute value.
+
+The nonlinear vector variable specified for $\vec{u}$ must be of the
+`RAVIART_THOMAS` family, as MOOSE currently does not compute the divergence for
+any other finite element family.
```
Comment (lindsayad) [inline framework/doc/content/source/postprocessors/ElementHDivSemiError.md:18]: I am changing this in #26477 so I would probably just remove this paragraph

File: framework/doc/content/source/postprocessors/ElementHDivSemiError.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+$\vec{f}$ must be a [Function](syntax/Functions/index.md) object defining both
+the `vectorValue` and `div` methods, e.g. `MooseParsedVectorFunction.md`.
```
Comment (lindsayad) [inline framework/doc/content/source/postprocessors/ElementHDivSemiError.md:21]:
```suggestion
the `vectorValue` and `div` methods, e.g. [MooseParsedVectorFunction.md].
```

File: framework/doc/content/source/postprocessors/ElementVectorL2Error.md [hunk 9]
```diff
@@ -4,12 +4,33 @@
...
+
+where $||\cdot||$ denotes the l2-norm in $\mathbb{R}^d$.
+
+The user can specify $\vec{u}$ with either a single nonlinear vector variable,
+or with up to three nonlinear scalar variables simultaneously, which would each
+typically represent the x, y, z components of the vector variable.
+
+Likewise, the user can also specify $\vec{f}$ with either a single
+[Function](syntax/Functions/index.md) object defining the `vectorValue` method,
+e.g. `MooseParsedVectorFunction.md`, or up to three such objects defining
```
Comment (lindsayad) [inline framework/doc/content/source/postprocessors/ElementVectorL2Error.md:22]:
```suggestion
e.g. [MooseParsedVectorFunction.md], or up to three such objects defining
```

File: framework/doc/content/source/postprocessors/ElementVectorL2Error.md [hunk 10]
```diff
@@ -4,12 +4,33 @@
...
+the `value` method, e.g. `MooseParsedFunction.md`.
```
Comment (lindsayad) [inline framework/doc/content/source/postprocessors/ElementVectorL2Error.md:23]:
```suggestion
the `value` method, e.g. [MooseParsedFunction.md].
```

File: framework/include/postprocessors/ElementHCurlSemiError.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+public:
+  static InputParameters validParams();
+
+  ElementHCurlSemiError(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpIntegral() override;
+
+  /// vector variable
+  VectorMooseVariable & _u_var;
```
Comment (lindsayad) [inline framework/include/postprocessors/ElementHCurlSemiError.h:30]:
```suggestion
  const VectorMooseVariable & _u_var;
```

File: framework/include/postprocessors/ElementHDivSemiError.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+public:
+  static InputParameters validParams();
+
+  ElementHDivSemiError(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpIntegral() override;
+
+  /// vector variable
+  VectorMooseVariable & _u_var;
```
Comment (lindsayad) [inline framework/include/postprocessors/ElementHDivSemiError.h:30]:
```suggestion
  const VectorMooseVariable & _u_var;
```

File: framework/src/postprocessors/ElementHCurlError.C [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+ElementHCurlError::ElementHCurlError(const InputParameters & parameters)
+  : ElementHCurlSemiError(parameters)
+{
+}
+
+Real
+ElementHCurlError::computeQpIntegral()
+{
+  return ElementVectorL2Error::computeQpIntegral() + ElementHCurlSemiError::computeQpIntegral();
```
Comment (lindsayad) [inline framework/src/postprocessors/ElementHCurlError.C:31]: big fan

File: framework/src/postprocessors/ElementVectorL2Error.C [hunk 14]
```diff
@@ -48,17 +71,14 @@ ElementVectorL2Error::getValue() const
...
 {
-  RealVectorValue sol_val(0.0, 0.0, 0.0);
-  RealVectorValue func_val(0.0, 0.0, 0.0);
-
-  sol_val(0) = _u[_qp];                          // required variable
-  func_val(0) = _funcx.value(_t, _q_point[_qp]); // required function
-
-  sol_val(1) = _v[_qp];
-  sol_val(2) = _w[_qp];
+  RealVectorValue func_val = isParamSetByUser("function")
```
Comment (lindsayad) [inline framework/src/postprocessors/ElementVectorL2Error.C:74]: I would populate a `_has_vector_function` or something like that in the constructor so we don't have to peruse the parameter database every time we hit a quadrature point

File: framework/src/postprocessors/ElementVectorL2Error.C [hunk 15]
```diff
@@ -48,17 +71,14 @@ ElementVectorL2Error::getValue() const
...
+                                 ? _func.vectorValue(_t, _q_point[_qp])
+                                 : RealVectorValue(_funcx.value(_t, _q_point[_qp]),
+                                                   _funcy.value(_t, _q_point[_qp]),
+                                                   _funcz.value(_t, _q_point[_qp]));
 
-  func_val(1) = _funcy.value(_t, _q_point[_qp]);
-  func_val(2) = _funcz.value(_t, _q_point[_qp]);
+  RealVectorValue sol_val =
+      isParamSetByUser("variable") ? _u[_qp] : RealVectorValue(_ux[_qp], _uy[_qp], _uz[_qp]);
```
Comment (lindsayad) [inline framework/src/postprocessors/ElementVectorL2Error.C:81]: similarly here

File: test/tests/kernels/vector_fe/tests [hunk 16]
```diff
@@ -8,6 +8,17 @@
...
+  [vector_fe_errors]
+    type = CSVDiff
+    input = vector_kernel.i
+    csvdiff = vector_kernel_out.csv
+    cli_args = 'Postprocessors/active="L2Error HCurlSemiError HCurlError" '
+               'Outputs/exodus=false Outputs/csv=true'
+    design = 'ElementVectorL2Error.md ElementHCurlSemiError.md ElementHCurlError.md'
+    issues = '#26394'
+    requirement = 'The system shall be able to estimate the L2 and H(curl) errors for a curl-curl '
+                  'problem solved using NEDELEC_ONE elements.'
```
Comment (lindsayad) [inline test/tests/kernels/vector_fe/tests:20]:
```suggestion
                  'problem solved using Nedelec elements of the first kind.'
```
You were probably going off the above bad example, but we try to keep implementation details out of requirements. From a thoretical point of view, we like to imagine that the requirement came before the implementation

File: test/tests/kernels/vector_fe/tests [hunk 17]
```diff
@@ -161,50 +172,122 @@
...
+                  'errors for a div-grad problem in H(div) solved using the '
+                  'following Raviart-Thomas elements: '
+    [hex27]
+      type = CSVDiff
+      input = coupled_electrostatics.i
+      csvdiff = coupled_electrostatics_hex27.csv
+      cli_args = 'Outputs/file_base=coupled_electrostatics_hex27 '
+                 'Postprocessors/active="PP L2Error HDivSemiError HDivError" '
+                 'Outputs/exodus=false Outputs/csv=true'
+      detail = 'HEX27, '
```
Comment (lindsayad) [inline test/tests/kernels/vector_fe/tests:237]: while appearing as implementation details, this is also a very concise way of describing exactly what kind of element it is, so I'm ok with this

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/doc/content/source/postprocessors/ElementHCurlSemiError.md (hunk 6 above)
Comment (nmnobre) [reply framework/doc/content/source/postprocessors/ElementHCurlSemiError.md]: Done.

File: framework/doc/content/source/postprocessors/ElementHDivSemiError.md (hunk 8 above)
Comment (nmnobre) [reply framework/doc/content/source/postprocessors/ElementHDivSemiError.md]: Done.

File: framework/doc/content/source/postprocessors/ElementVectorL2Error.md (hunk 9 above)
Comment (nmnobre) [reply framework/doc/content/source/postprocessors/ElementVectorL2Error.md]: Done.

File: framework/doc/content/source/postprocessors/ElementVectorL2Error.md (hunk 10 above)
Comment (nmnobre) [reply framework/doc/content/source/postprocessors/ElementVectorL2Error.md]: Done.

File: test/tests/kernels/vector_fe/tests (hunk 16 above)
Comment (nmnobre) [reply test/tests/kernels/vector_fe/tests]: Done, and also fixed the "bad" examples in the same file in a separate commit.

File: framework/include/postprocessors/ElementHCurlSemiError.h (hunk 11 above)
Comment (nmnobre) [reply framework/include/postprocessors/ElementHCurlSemiError.h]: Done.

File: framework/include/postprocessors/ElementHDivSemiError.h (hunk 12 above)
Comment (nmnobre) [reply framework/include/postprocessors/ElementHDivSemiError.h]: Done, and also fixed a few other occurrences elsewhere in a different commit.

File: framework/src/postprocessors/ElementHCurlError.C (hunk 13 above)
Comment (nmnobre) [reply framework/src/postprocessors/ElementHCurlError.C:31]: 🤗

File: framework/doc/content/source/postprocessors/ElementHDivSemiError.md (hunk 7 above)
Comment (nmnobre) [reply framework/doc/content/source/postprocessors/ElementHDivSemiError.md:18]: Correct me if I'm wrong.
You are changing things in #26477 so the decision to know if divergences are needed or not will depend simply on a request for that quantity by the user? That means, divergences will now be available for all vector fe families (except for `NEDELEC_ONE` for which the H(curl) Piola mapping is missing libMesh)?
I see you don't intend to make the same change for curls, so should I keep the disclaimer in `ElementHCurlSemiError.md`?

File: framework/src/postprocessors/ElementVectorL2Error.C (hunk 14 above)
Comment (nmnobre) [reply framework/src/postprocessors/ElementVectorL2Error.C]: Good idea, done.

File: framework/src/postprocessors/ElementVectorL2Error.C (hunk 15 above)
Comment (nmnobre) [reply framework/src/postprocessors/ElementVectorL2Error.C]: Done.

**Comment by nmnobre** [comment]
All done, I think. :)
I can squash (or move elsewhere) the last three commits which have changes similar to those you have suggested but for (parts of) files I didn't otherwise touch in this PR.

File: framework/doc/content/source/postprocessors/ElementHDivSemiError.md (hunk 7 above)
Comment (lindsayad) [reply framework/doc/content/source/postprocessors/ElementHDivSemiError.md]: Correct on all counts

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
Thank you @nmnobre !

**Pull request merged**